src/Microsoft.AspNetCore.Routing/EndpointMiddleware.cs

```diff
@@ -32,26 +32,18 @@ public EndpointMiddleware(ILogger<EndpointMiddleware> logger, RequestDelegate ne
 
         public async Task Invoke(HttpContext httpContext)
         {
-            var feature = httpContext.Features.Get<IEndpointFeature>();
-            if (feature == null)
+            var endpoint = httpContext.Features.Get<IEndpointFeature>()?.Endpoint;
+            if (endpoint?.RequestDelegate != null)
             {
-                var message = $"Unable to execute an endpoint because the {nameof(EndpointRoutingMiddleware)} was not run for this request. " +
```

Comment:
With hosting taking the responsibility of invoking the endpoint in 3.0 we won't need this anymore.

Reply:
Hmmm, tbh it depends on what we want from this middleware.

Is the purpose of it:
1. I have an endpoint, stop now and don't execute anymore middlware (and if there isn't then throw an exception)
2. Stop if we have an endpoint, otherwise keep going

IMO I think (1) is what people would expect as a default. Maybe it should be an option?

Reply:
So in 3.0 I don't think there's a purpose for this middleware at all - because this becomes a built in behavior - so the question is academic. Does that match your understanding?

I would definitely expect 2. Middleware generally don't have preconditions to validate, they no op if there's nothing to do.

Reply:
In my mind this is like a return statement from a method. Even if a method implicitly returned a value at the end, I still then there are cases where you might want to explicitly return earlier.

> I would definitely expect 2. Middleware generally don't have preconditions to validate, they no op if there's nothing to do.

Ok, but I wonder if this middleware should be stricter since it is an endpoint rather a step along a pipline.

Maybe I'm overthinking this 😬

Reply:
I don't think you're overthinking, I think you're ascribing some personality characteristics to middleware that they don't/shouldn't have. In general middleware are designed for 
 maximum composability.

Put another way, if you don't want to handle something, maybe someone else does. It's much better to defer the responsiblity to decide how to response than it is to report a 404 or a 500 yourself. It's the chain of responsibility pattern.

Reply:
Ok. If someone does want to force a hard exit then they have tools to do so easily without a helper middleware.